pallets/automation-time/src/migrations.rs

```diff
@@ -1,72 +1,327 @@
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+			let pallet_prefix: &[u8] = b"AutomationTime";
+
+			// Get count of tasks
+			let tasks_prefix: &[u8] = b"Tasks";
+			let pre_task_count =
+				storage_key_iter::<T::Hash, Task<T>, Twox64Concat>(pallet_prefix, tasks_prefix)
+					.collect::<Vec<_>>()
+					.len() as u32;
+			Self::set_temp_storage::<u32>(pre_task_count, "pre_task_count");
+
+			// Get count per scheduled time
+			let scheduled_prefix: &[u8] = b"ScheduledTasks";
+			let mut pre_scheduled_count: Vec<(UnixTime, u32)> = vec![];
+			storage_key_iter::<u64, BoundedVec<T::Hash, T::MaxTasksPerSlot>, Twox64Concat>(
+				pallet_prefix,
+				scheduled_prefix,
+			)
+			.for_each(|(time, task_ids)| {
+				pre_scheduled_count.push((time, task_ids.len() as u32));
+			});
+			Self::set_temp_storage::<Vec<(UnixTime, u32)>>(
+				pre_scheduled_count,
+				"pre_scheduled_count",
+			);
+
+			// Get count of tasks in task queue
+			let task_queue_prefix: &[u8] = b"TaskQueue";
+			let pre_task_queue_count =
+				get_storage_value::<Vec<TaskId<T>>>(pallet_prefix, task_queue_prefix, &[])
+					.unwrap_or(vec![])
+					.len() as u32;
+			Self::set_temp_storage::<u32>(pre_task_queue_count, "pre_task_queue_count");
+
+			// Get count of tasks in missed queue
+			let missed_queue_prefix: &[u8] = b"MissedQueue";
+			let pre_missed_queue_count =
+				get_storage_value::<Vec<MissedTask<T>>>(pallet_prefix, missed_queue_prefix, &[])
+					.unwrap_or(vec![])
+					.len() as u32;
+			Self::set_temp_storage::<u32>(pre_missed_queue_count, "pre_missed_queue_count");
+
+			log::info!(
+				target: "automation-time",
+				"migration: AutomationTime storage version v3 PRE migration checks succesful!"
+			);
+
+			Ok(())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			log::info!(target: "automation-time", "Migrating automation-time v3");
+			let pallet_prefix: &[u8] = b"AutomationTime";
+
+			// Move all tasks from Tasks to AccountTasks
+			let old_tasks_prefix: &[u8] = b"Tasks";
+			let old_tasks =
+				storage_key_iter::<T::Hash, Task<T>, Twox64Concat>(pallet_prefix, old_tasks_prefix)
+					.drain()
+					.collect::<btree_map::BTreeMap<T::Hash, Task<T>>>();
+			old_tasks.iter().for_each(|(task_id, task)| {
+				AccountTasks::<T>::insert(task.owner_id.clone(), task_id, task.clone());
+			});
+			let task_migrated = old_tasks.len() as u64;
+
+			// Move all tasks from ScheduledTasks to ScheduledTasksV2
+			let mut times_migrated = 0u64;
+			let old_scheduled_prefix: &[u8] = b"ScheduledTasks";
+			storage_key_iter::<u64, BoundedVec<T::Hash, T::MaxTasksPerSlot>, Twox64Concat>(
+				pallet_prefix,
+				old_scheduled_prefix,
+			)
+			.drain()
+			.for_each(|(time, task_ids)| {
+				let new_task_ids = task_ids
+					.into_iter()
+					.filter_map(|task_id| {
+						if let Some(task) = old_tasks.get(&task_id) {
```

Comment:
wait how does this work? isn't old_tasks an arry?

Reply:
it's a btree